File: backend/app/application/hwpx/ast/html_table_to_ast.py
"""HTML table → 표 AST."""

from __future__ import annotations

import re
import uuid
from typing import Any

from app.application.hwpx.ast.normalize_table_grid import normalize_table_grid
from app.application.hwpx.ast.types import HWP_TABLE_WIDTH_HWPUNIT
from app.application.hwpx.encoding import sanitize_hwpx_text


def _ast_id(prefix: str) -> str:
    return f"{prefix}-{uuid.uuid4()}"


def _cell_text(el) -> str:
    raw = " ".join((el.text_content() or "").split())
    return sanitize_hwpx_text(raw) or " "


def _parse_css_length(raw: str | None) -> int | None:
    if not raw:
        return None
    px = re.match(r"^([\d.]+)px$", raw.strip(), re.I)
    if px:
        return int(float(px.group(1)))
    pct = re.match(r"^([\d.]+)%$", raw.strip(), re.I)
    if pct:
        return int(HWP_TABLE_WIDTH_HWPUNIT * float(pct.group(1)) / 100)
    return None


def _extract_column_widths(table_el, col_count: int) -> list[int]:
    widths: list[int] = []
    for col in table_el.xpath("./colgroup/col"):
        w = _parse_css_length(col.get("data-col-width"))
        if w is None:
            style = col.get("style") or ""
            m = re.search(r"width:\s*([^;]+)", style, re.I)
            w = _parse_css_length(m.group(1).strip()) if m else None
        widths.append(w or 0)

    if len(widths) >= col_count and all(w > 0 for w in widths[:col_count]):
        return widths[:col_count]

    total = HWP_TABLE_WIDTH_HWPUNIT
    base = total // col_count
    return [base] * (col_count - 1) + [total - base * (col_count - 1)]


def _parse_font_size_px(raw: str | None) -> int | None:
    if not raw:
        return None
    m = re.match(r"^([\d.]+)\s*px$", raw.strip(), re.I)
    if not m:
        return None
    px = int(round(float(m.group(1))))
    return px if px > 0 else None


def _cell_font_size_px(cell_el) -> int | None:
    """셀 대표 글자 크기(px) — 셀 자체 또는 첫 명시 스팬 기준."""
    own = _parse_font_size_px(cell_el.get("data-bp-fz"))
    if own:
        return own
    style = cell_el.get("style") or ""
    m = re.search(r"font-size:\s*([^;]+)", style, re.I)
    own = _parse_font_size_px(m.group(1).strip()) if m else None
    if own:
        return own
    for marked in cell_el.xpath(".//*[@data-bp-fz]"):
        fz = _parse_font_size_px(marked.get("data-bp-fz"))
        if fz:
            return fz
    for styled in cell_el.xpath('.//*[contains(@style, "font-size")]'):
        sm = re.search(r"font-size:\s*([^;]+)", styled.get("style") or "", re.I)
        fz = _parse_font_size_px(sm.group(1).strip()) if sm else None
        if fz:
            return fz
    return None


def _html_cell_to_ast(cell_el) -> dict[str, Any]:
    tag = (cell_el.tag or "").lower()
    style = cell_el.get("style") or ""
    bg = cell_el.get("data-bg")
    if not bg:
        m = re.search(r"background-color:\s*([^;]+)", style, re.I)
        bg = m.group(1).strip() if m else None

    valign = "top"
    vm = re.search(r"vertical-align:\s*(\w+)", style, re.I)
    if vm and vm.group(1) in ("top", "middle", "bottom"):
        valign = vm.group(1)

    text = _cell_text(cell_el)
    return {
        "type": "tableCell",
        "id": cell_el.get("data-ast-id") or _ast_id("tc"),
        "rowSpan": int(cell_el.get("rowspan") or 1),
        "colSpan": int(cell_el.get("colspan") or 1),
        "grid": {"row": 0, "col": 0},
        "style": {
            "backgroundColor": bg,
            "verticalAlign": valign,
            "header": tag == "th",
            "fontSizePx": _cell_font_size_px(cell_el),
        },
        "content": [
            {
                "type": "paragraph",
                "content": [{"type": "text", "text": text}],
            }
        ]
        if text.strip()
        else [],
        "hwp": {},
    }


def html_table_element_to_ast(table_el) -> dict[str, Any]:
    rows: list[dict[str, Any]] = []
    for tr in table_el.xpath("./tbody/tr | ./tr"):
        cells = [_html_cell_to_ast(c) for c in tr.xpath("./td|./th")]
        if cells:
            rows.append(
                {
                    "type": "tableRow",
                    "id": tr.get("data-ast-id") or _ast_id("tr"),
                    "cells": cells,
                }
            )

    col_spans = [
        sum(int(c.get("colSpan") or 1) for c in row.get("cells") or [])
        for row in rows
    ]
    col_count = max(1, max(col_spans)) if col_spans else 1

    if not rows:
        rows = [
            {
                "type": "tableRow",
                "id": _ast_id("tr"),
                "cells": [
                    {
                        "type": "tableCell",
                        "id": _ast_id("tc"),
                        "rowSpan": 1,
                        "colSpan": 1,
                        "grid": {"row": 0, "col": 0},
                        "style": {"verticalAlign": "top"},
                        "content": [
                            {
                                "type": "paragraph",
                                "content": [{"type": "text", "text": " "}],
                            }
                        ],
                    }
                ],
            }
        ]

    return normalize_table_grid(
        {
            "type": "table",
            "id": table_el.get("data-ast-id") or _ast_id("tbl"),
            "columns": [{"width": w} for w in _extract_column_widths(table_el, col_count)],
            "rows": rows,
            "hwp": {},
        }
    )


def paragraph_text_from_blocks(blocks: list[dict[str, Any]]) -> str:
    parts: list[str] = []
    for block in blocks:
        if block.get("type") != "paragraph":
            continue
        for item in block.get("content") or []:
            if item.get("type") == "text":
                text = sanitize_hwpx_text(str(item.get("text") or "")).strip()
                if text:
                    parts.append(text)
    return " ".join(parts)

File: frontend-next/lib/hwp-ast/htmlTableToAst.ts
import { hwpAstId } from "@/lib/hwp-ast/ids"
import { normalizeTableGrid } from "@/lib/hwp-ast/normalizeTableGrid"
import type {
  HwpBlock,
  HwpParagraphBlock,
  HwpTable,
  HwpTableCell,
  HwpTableRow,
} from "@/lib/hwp-ast/types"
import { HWP_TABLE_WIDTH_HWPUNIT } from "@/lib/hwp-ast/types"
import { sanitizeHwpxText } from "@/lib/hwpx/hwpx-encoding"

function cellText(el: Element): string {
  return sanitizeHwpxText((el.textContent ?? "").replace(/\s+/g, " ").trim() || " ")
}

function parseFontSizePx(raw: string | null | undefined): number | null {
  if (!raw) return null
  const m = /^([\d.]+)\s*px$/i.exec(raw.trim())
  if (!m) return null
  const px = Math.round(Number(m[1]))
  return Number.isFinite(px) && px > 0 ? px : null
}

/** 셀의 대표 글자 크기(px) — 셀 자체 또는 첫 번째 명시 스팬 기준 */
function cellFontSizePx(cellEl: HTMLTableCellElement): number | null {
  const own =
    parseFontSizePx(cellEl.getAttribute("data-bp-fz")) ??
    parseFontSizePx(cellEl.style.fontSize)
  if (own) return own

  const marked = cellEl.querySelector<HTMLElement>("[data-bp-fz]")
  if (marked) {
    const fromAttr = parseFontSizePx(marked.getAttribute("data-bp-fz"))
    if (fromAttr) return fromAttr
  }

  const styled = cellEl.querySelector<HTMLElement>('[style*="font-size"]')
  if (styled) {
    const fromStyle = parseFontSizePx(styled.style.fontSize)
    if (fromStyle) return fromStyle
  }

  return null
}

function parseCssLength(raw: string | null | undefined): number | null {
  if (!raw) return null
  const px = raw.match(/^([\d.]+)px$/i)
  if (px) return Math.round(Number(px[1]))
  const pct = raw.match(/^([\d.]+)%$/i)
  if (pct) {
    return Math.round((HWP_TABLE_WIDTH_HWPUNIT * Number(pct[1])) / 100)
  }
  return null
}

function extractColumnWidths(tableEl: HTMLTableElement, colCount: number): number[] {
  const fromColgroup: number[] = []
  const colgroup = tableEl.querySelector("colgroup")
  if (colgroup) {
    for (const col of Array.from(colgroup.querySelectorAll("col"))) {
      const w =
        parseCssLength(col.getAttribute("data-col-width")) ??
        parseCssLength(col.style.width) ??
        parseCssLength(col.getAttribute("width"))
      fromColgroup.push(w ?? 0)
    }
  }

  if (fromColgroup.length >= colCount && fromColgroup.every((w) => w > 0)) {
    return fromColgroup.slice(0, colCount)
  }

  const total = HWP_TABLE_WIDTH_HWPUNIT
  const base = Math.floor(total / colCount)
  return Array.from({ length: colCount }, (_, i) =>
    i === colCount - 1 ? total - base * (colCount - 1) : base,
  )
}

function htmlCellToAst(cellEl: HTMLTableCellElement): HwpTableCell {
  const bg =
    cellEl.getAttribute("data-bg") ??
    cellEl.style.backgroundColor ??
    null

  const verticalAlign =
    (cellEl.style.verticalAlign as "top" | "middle" | "bottom" | "") || "top"

  return {
    type: "tableCell",
    id: cellEl.getAttribute("data-ast-id") ?? hwpAstId("tc"),
    rowSpan: cellEl.rowSpan || 1,
    colSpan: cellEl.colSpan || 1,
    grid: { row: 0, col: 0 },
    style: {
      backgroundColor: bg,
      verticalAlign: verticalAlign || "top",
      header: cellEl.tagName === "TH",
      fontSizePx: cellFontSizePx(cellEl),
    },
    content: cellText(cellEl)
      ? [{ type: "paragraph", content: [{ type: "text", text: cellText(cellEl) }] }]
      : [],
    hwp: {},
  }
}

export function htmlTableToAst(tableEl: HTMLTableElement): HwpTable {
  const rows: HwpTableRow[] = []

  for (let r = 0; r < tableEl.rows.length; r++) {
    const tr = tableEl.rows[r]
    const cells: HwpTableCell[] = []
    for (let c = 0; c < tr.cells.length; c++) {
      cells.push(htmlCellToAst(tr.cells[c]))
    }
    if (cells.length > 0) {
      rows.push({
        type: "tableRow",
        id: tr.getAttribute("data-ast-id") ?? hwpAstId("tr"),
        cells,
      })
    }
  }

  const colCount = Math.max(
    1,
    ...rows.map((row) => row.cells.reduce((sum, cell) => sum + cell.colSpan, 0)),
  )

  return normalizeTableGrid({
    type: "table",
    id: tableEl.getAttribute("data-ast-id") ?? hwpAstId("tbl"),
    columns: extractColumnWidths(tableEl, colCount).map((width) => ({ width })),
    rows: rows.length > 0 ? rows : [{
      type: "tableRow",
      id: hwpAstId("tr"),
      cells: [{
        type: "tableCell",
        id: hwpAstId("tc"),
        rowSpan: 1,
        colSpan: 1,
        grid: { row: 0, col: 0 },
        style: { verticalAlign: "top" },
        content: [{ type: "paragraph", content: [{ type: "text", text: " " }] }],
      }],
    }],
    hwp: {},
  })
}

export type HtmlHwpxBlocks = {
  paragraphs: Array<{ text: string; variant?: "title" | "heading" | "body" }>
  tables: HwpTable[]
}

function pushParagraph(
  blocks: HtmlHwpxBlocks,
  text: string,
  variant?: "title" | "heading" | "body",
) {
  const t = text.trim()
  if (t) blocks.paragraphs.push({ text: t, variant })
}

/** 리치텍스트 HTML → 문단 + 표 AST */
export function htmlToHwpBlocks(html: string): HtmlHwpxBlocks {
  const blocks: HtmlHwpxBlocks = { paragraphs: [], tables: [] }
  if (!html?.trim()) return blocks

  if (typeof document === "undefined") {
    return blocks
  }

  const root = document.createElement("div")
  root.innerHTML = html

  const walk = (parent: Element) => {
    for (const node of Array.from(parent.childNodes)) {
      if (node.nodeType === Node.TEXT_NODE) {
        const t = node.textContent?.trim()
        if (t) pushParagraph(blocks, t, "body")
        continue
      }
      if (node.nodeType !== Node.ELEMENT_NODE) continue
      const el = node as HTMLElement
      const tag = el.tagName.toLowerCase()

      if (tag === "table") {
        blocks.tables.push(htmlTableToAst(el as HTMLTableElement))
        continue
      }
      if (tag === "h1" || tag === "h2" || el.classList.contains("doc-chapter")) {
        pushParagraph(blocks, cellText(el), "heading")
        continue
      }
      if (
        tag === "h3" ||
        tag === "h4" ||
        el.classList.contains("doc-section") ||
        el.classList.contains("doc-section-plain")
      ) {
        pushParagraph(blocks, cellText(el), "heading")
        continue
      }
      if (tag === "p" || tag === "li") {
        pushParagraph(blocks, cellText(el), "body")
        continue
      }
      if (tag === "img") {
        const alt = sanitizeHwpxText(el.getAttribute("alt") || "") || "[이미지]"
        pushParagraph(blocks, alt, "body")
        continue
      }
      if (tag === "br") continue
      if (["div", "ul", "ol", "tbody", "thead"].includes(tag)) {
        walk(el)
        continue
      }
      pushParagraph(blocks, cellText(el), "body")
    }
  }

  walk(root)
  return blocks
}

export function paragraphTextFromBlocks(blocks: HwpBlock[]): string {
  const parts: string[] = []
  for (const block of blocks) {
    if (block.type !== "paragraph") continue
    for (const run of block.content) {
      if (run.type === "text" && run.text.trim()) {
        parts.push(run.text.trim())
      }
    }
  }
  return parts.join(" ")
}
